apps/emqx_durable_storage/src/emqx_ds_bitmask_keymapper.erl

```diff
@@ -0,0 +1,746 @@
+) ->
+    NDim = length(DimSizeof),
+    %% Transform "symbolic" inequations to ranges:
+    Filter1 = inequations_to_ranges(KeyMapper, Filter0),
+    {Bitmask, Bitfilter} = make_bitfilter(KeyMapper, Filter1),
+    %% Calculate maximum source offset as per bitsource specification:
+    MaxOffset = lists:foldl(
+        fun({Dim, Offset, _Size}, Acc) ->
+            maps:update_with(Dim, fun(OldVal) -> max(OldVal, Offset) end, 0, Acc)
+        end,
+        #{},
+        Schema
+    ),
+    %% Adjust minimum and maximum values for each interval like this:
+    %%
+    %% Min: 110100|101011 -> 110100|00000
+    %% Max: 110101|001011 -> 110101|11111
+    %%            ^
+    %%            |
+    %%       max offset
+    %%
+    %% This is needed so when we increment the vector, we always scan
+    %% the full range of least significant bits.
+    Filter2 = lists:map(
+        fun
+            ({{Val, Val}, _Dim}) ->
+                {Val, Val};
+            ({{Min0, Max0}, Dim}) ->
+                Offset = maps:get(Dim, MaxOffset, 0),
+                %% Set least significant bits of Min to 0:
+                Min = (Min0 bsr Offset) bsl Offset,
+                %% Set least significant bits of Max to 1:
+                Max = Max0 bor ones(Offset),
+                {Min, Max}
+        end,
+        lists:zip(Filter1, lists:seq(1, NDim))
+    ),
+    %% Project the vector into "bitsource coordinate system":
+    {_, Filter} = fold_bitsources(
+        fun(DstOffset, {Dim, SrcOffset, Size}, Acc) ->
+            {Min0, Max0} = lists:nth(Dim, Filter2),
+            Min = (Min0 bsr SrcOffset) band ones(Size),
+            Max = (Max0 bsr SrcOffset) band ones(Size),
+            Action = #filter_scan_action{
+                offset = DstOffset,
+                size = Size,
+                min = Min,
+                max = Max
+            },
+            [Action | Acc]
+        end,
+        [],
+        Schema
+    ),
+    Ranges = array:from_list(lists:reverse(Filter)),
+    %% Compute estimated upper and lower bounds of a _continous_
+    %% interval where all keys lie:
+    case Filter of
+        [] ->
+            RangeMin = 0,
+            RangeMax = 0;
+        [#filter_scan_action{offset = MSBOffset, min = MSBMin, max = MSBMax} | _] ->
+            RangeMin = MSBMin bsl MSBOffset,
+            RangeMax = MSBMax bsl MSBOffset bor ones(MSBOffset)
+    end,
+    %% Final value
+    #filter{
+        size = TotalSize,
+        bitmask = Bitmask,
+        bitfilter = Bitfilter,
+        bitsource_ranges = Ranges,
+        range_min = RangeMin,
+        range_max = RangeMax
+    }.
+
+-spec ratchet(filter(), key()) -> key() | overflow.
+ratchet(#filter{bitsource_ranges = Ranges, range_max = Max}, Key) when Key =< Max ->
+    NDim = array:size(Ranges),
+    case ratchet_scan(Ranges, NDim, Key, 0, _Pivot = {-1, 0}, _Carry = 0) of
+        overflow ->
+            overflow;
+        {Pivot, Increment} ->
```

Comment:
Small nit: it's slightly confusing that `Pivot` when calling seems to be `{Dim, Increment}`, but here `Pivot` becomes `Dim` (assuming it's really a dimension/component index).

Reply:
Thanks, fixed.